## Problem2

This dataset contains information of users in a social network. This social network has several business clients which can post ads on it. One of the clients has a car company which has just launched a luxury SUV for a ridiculous price. Build a Bernoulli Naïve Bayes model using this dataset and classify which of the users of the social network are going to purchase this luxury SUV. 1 implies that there was a purchase and 0 implies there wasn’t a purchase.   

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [3]:
sns.set()

In [4]:
# import Dataset 
#sal_train = pd.read_csv("SalaryData_Train.csv", encoding = 'utf8')
car = pd.read_csv("NB_Car_Ad.csv")

In [5]:
car.shape

(400, 5)

In [6]:
car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   User ID          400 non-null    int64 
 1   Gender           400 non-null    object
 2   Age              400 non-null    int64 
 3   EstimatedSalary  400 non-null    int64 
 4   Purchased        400 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 15.8+ KB


In [9]:
car.head()

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

In [7]:
car.isnull().sum()

User ID            0
Gender             0
Age                0
EstimatedSalary    0
Purchased          0
dtype: int64

In [32]:
#User ID
car['User ID'].nunique()

400

In [33]:
#we can drop User ID as it provides no valuable info
car.drop(['User ID'],axis=1,inplace=True)

In [34]:
car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Gender           400 non-null    object
 1   Age              400 non-null    int64 
 2   EstimatedSalary  400 non-null    int64 
 3   Purchased        400 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 12.6+ KB


In [35]:
#check for duplicates
car.duplicated().sum()

20

In [38]:
#drop duplicates
car.drop_duplicates(inplace=True)

In [39]:
car.duplicated().any()

False

In [40]:
car.iloc[:,:]

Gender  Age  EstimatedSalary  Purchased
0      Male   19            19000          0
1      Male   35            20000          0
2    Female   26            43000          0
3    Female   27            57000          0
4      Male   19            76000          0
..      ...  ...              ...        ...
395  Female   46            41000          1
396    Male   51            23000          1
397  Female   50            20000          1
398    Male   36            33000          0
399  Female   49            36000          1

[380 rows x 4 columns]

In [42]:
X = np.array(car.iloc[:,:3]) # Predictors 
Y = np.array(car.iloc[:,3]) # Target

In [43]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2)

In [44]:
X_train.shape

(304, 3)

In [45]:
X_test.shape

(76, 3)

In [46]:
Y_train.shape

(304,)

In [47]:
Y_test.shape

(76,)

In [49]:
X_train[1,:]

array(['Female', 42, 80000], dtype=object)

In [85]:
Y_train

array([1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1,
       0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1,

In [74]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.compose import ColumnTransformer
from sklearn.compose import make_column_transformer
from sklearn.pipeline import Pipeline

nominal_features=[0]
numeric_features = [1,2]


############# numeric_transformer ####################
numeric_transformer = make_column_transformer(
    (MinMaxScaler(), numeric_features),    
)

############# categorical_transformer ####################

categorical_transformer = make_column_transformer(
    (OneHotEncoder(drop='first'), nominal_features),
)


preprocessor = ColumnTransformer(
    transformers=[
        ("categorical", categorical_transformer, [0,1,2]),
        ("numeric", numeric_transformer, [0,1,2]),
    ]
)



In [75]:
# We will later Append classifier to preprocessing pipeline.
pipe = Pipeline(
    steps=[("preprocessor", preprocessor)]
)

In [76]:
# Apply the tranformation to train data
X_train_transformed = pipe.fit_transform(X_train) 

In [79]:
X_train_transformed[:2,0:]

array([[0.        , 0.85714286, 0.65925926],
       [0.        , 0.57142857, 0.48148148]])

In [80]:
# Apply the tranformation to test data
X_test_transformed = pipe.fit_transform(X_test) 

In [81]:
X_train_transformed.shape

(304, 3)

In [82]:
X_test_transformed.shape

(76, 3)

In [88]:
#Train Data
X_train = X_train_transformed[: , :] # Predictors 

In [89]:
#Test Data
X_test = X_test_transformed[: , :] # Predictors 

In [90]:
# Preparing a naive bayes model on training data set 
from sklearn.naive_bayes import BernoulliNB

# Bernoulli Naive Bayes
classifier_bb = BernoulliNB()
classifier_bb.fit(X_train,Y_train,sample_weight=None)

BernoulliNB()

In [107]:
# Evaluation on Test Data
test_pred_m = classifier_bb.predict(X_test)
accuracy_test_m = np.mean(test_pred_m == Y_test)
accuracy_test_m

0.5789473684210527

In [108]:
#accuracy in test data
from sklearn.metrics import accuracy_score
accuracy_score(test_pred_m, Y_test) 

0.5789473684210527

In [109]:
pd.crosstab(test_pred_m, Y_test)

col_0   0   1
row_0        
0      44  32

In [102]:
#accuracy in training data
test_pred_m = classifier_bb.predict(X_train)
accuracy_score(test_pred_m, Y_train) 

0.6447368421052632

In [112]:
# Multinomial Naive Bayes changing default alpha for laplace smoothing
# if alpha = 0 then no smoothing is applied and the default alpha parameter is 1
# the smoothing process mainly solves the emergence of zero probability problem in the dataset.
classifier_bb_lap = BernoulliNB(alpha =1)
classifier_bb_lap.fit(X_train, Y_train)

# Evaluation on Test Data after applying laplace
test_pred_lap = classifier_bb_lap.predict(X_test)
accuracy_test_lap = np.mean(test_pred_lap == Y_test)
accuracy_test_lap


0.5789473684210527

with laplace smoothening with alpha 1 we get close to 58% accuracy